common/etc/nginx/include/s3gateway.js

```diff
@@ -643,10 +690,120 @@ function _require_env_var(envVarName) {
+ *
+ * To make sure we always refresh the credentials instead of retrieving the same again, keep credentials until 4:30 minutes
+ * before they really expire.
+ *
+ * @type {number}
+ */
+var maxValidityOffsetMs = 4.5 * 60 * 100;
+
+async function fetchCredentials(r) {
+    var current = readCredentials();
+    if (current) {
+        var exp = new Date(current.expiration).getTime() - maxValidityOffsetMs;
+        if (now.getTime() < exp) {
+            r.return(200);
+            return;
+        }
+    }
+    _debug_log(r, 'Cached credentials are expired or not present, requesting new ones');
+
+    var credentials;
+    if (process.env['S3_ACCESS_KEY_ID']) {
+        credentials = {
+            accessKeyId: process.env['S3_ACCESS_KEY_ID'],
+            secretAccessKey: process.env['S3_SECRET_KEY'],
+            sessionToken: null,
+        };
+    } else if (process.env['AWS_CONTAINER_CREDENTIALS_RELATIVE_URI']) {
+        var uri = ecsCredentialsBaseUri + process.env['AWS_CONTAINER_CREDENTIALS_RELATIVE_URI']
+        try {
+            credentials = await _fetchEcsRoleCredentials(uri);
+        } catch (e) {
+            _debug_log(r, 'Could not load ECS task role credentials: ' + JSON.stringify(e));
+            r.return(500);
+            return;
+        }
+    } else {
+        try {
+            credentials = await _fetchEC2RoleCredentials();
+        } catch (e) {
+            _debug_log(r, 'Could not load EC2 task role credentials: ' + JSON.stringify(e));
+            r.return(500);
+            return;
+        }
+    }
+    try {
+        fs.writeFileSync(_credentialsTempFile(), JSON.stringify(credentials));
+    } catch (e) {
+        _debug_log(r, 'Could not write credentials file: ' + JSON.stringify(e));
+        r.return(500);
+        return;
+    }
+    r.return(200);
+}
+
+async function _fetchEcsRoleCredentials(credentialsUri) {
+    var resp = await ngx.fetch(credentialsUri);
+    if (!resp.ok) {
+        throw 'Credentials endpoint response was not ok.';
+    }
+    var creds = await resp.json();
+
+    return {
+        accessKeyId: creds.AccessKeyId,
+        secretAccessKey: creds.SecretAccessKey,
+        sessionToken: creds.Token,
+        expiration: creds.Expiration,
+    };
+}
+
+async function _fetchEC2RoleCredentials() {
+    var tokenResp = await ngx.fetch(ec2ImdsTokenEndpoint, {
+        headers: {
+            'x-aws-ec2-metadata-token-ttl-seconds': '21600',
+        },
+        method: 'PUT',
+    });
+    var token = await tokenResp.text();
+    var resp = await ngx.fetch(ec2ImdsSecurityCredentialsEndpoint, {
+        headers: {
+            'x-aws-ec2-metadata-token': token,
+        },
+    });
+    // This _might_ get multiple possible roles in other scenarios, however, EC2 supports attaching one role only.
+    // It should therefore be safe to take the whole output, even given IMDS _might_ (?) be able to return multiple
+    // roles.
+    var credName = resp.text();
```

Comment:
Shouldn't we be doing an `await` here? Like:
```
var credName = await resp.text();
```

Reply:
Embarrassing, you're absolutely right, my fault 👍